### PR TITLE
fix: subuser and api key now mostly work

### PR DESCRIPTION
**Pull request by timja**
Still debugging:

```
Error: failed creating apiKey, status: 500, response: {"errors":[{"field":null,"message":"internal server error"}]}

  on main.tf line 45, in resource "sendgrid_api_key" "api_key_sub_user_1":
  45: resource "sendgrid_api_key" "api_key_sub_user_1" {
```

### CODE REVIEW

**Comment by Trois-Six** [comment]
OK I understand what you want to do and sorry to have removed the special case for 404. If we follow that path, we should also handle the retry for api_keys and all other resources.

**Comment by timja** [comment]
@Trois-Six I partially reverted one of your fixes for 2fa

I'm using an account via a reseller (Azure) and adding that scope causes send grid to respond with an internal server error.

I'm guessing it didn't work for you without adding the scope? Or any other reason you force added it

**Comment by timja** [comment]
> we should also handle the retry for api_keys and all other resources; as you did for subUsers.

Yes that needs adding too:

> Error: failed creating apiKey, status: 429, response: {"errors":[{"field":null,"message":"too many requests"}]}

**Comment by Trois-Six** [comment]
> @Trois-Six I partially reverted one of your fixes for 2fa ([80d523e](https://github.com/Trois-Six/terraform-provider-sendgrid/commit/80d523e88b92eb63efb3acea059b9d21128ef66d))
> 
> I'm using an account via a reseller (Azure) and adding that scope causes send grid to respond with an internal server error.
> 
> I'm guessing it didn't work for you without adding the scope? Or any other reason you force added it

It's not that it doesn't work, but that scope is automatically added even if I do not set it. If I don't do that, each time I run terraform I have a change.

**Comment by Trois-Six** [comment]
I don't know how to handle the linter error here: https://github.com/Trois-Six/terraform-provider-sendgrid/pull/6/files#diff-4b6d2ae9db41835ea2a1cdb747d065675c0596695ad0f7a5c5012c9b0ce89e93R34.

File: sendgrid/resource_sendgrid_subuser.go [hunk 1]
```diff
@@ -31,7 +30,9 @@ import (
 	sendgrid "github.com/trois-six/terraform-provider-sendgrid/sdk"
 )
 
-var ErrSubUserNotFound = errors.New("subUser wasn't found")
+func subUserNotFound(name string) error {
+	return fmt.Errorf("subUser %s wasn't found", name)
```
Comment (Trois-Six) [inline sendgrid/resource_sendgrid_subuser.go:34]: that creates a linter warning (err113).

File: sdk/api_key.go [hunk 2]
```diff
@@ -70,21 +96,39 @@ func (c *Client) UpdateAPIKey(id, name string, scopes []string) (*APIKey, error)
...
+	if err != nil {
+		return false, RequestError{
+			StatusCode: http.StatusInternalServerError,
+			Err:        fmt.Errorf("failed deleting API key: %w", err),
+		}
 	}
 
-	if _, statusCode, err := c.Get("DELETE", "/api_keys/"+id); statusCode > 299 || err != nil {
-		return false, fmt.Errorf("failed deleting API key: %w", err)
+	if statusCode >= 300 && statusCode != 404 { // ignore not found
```
Comment (Trois-Six) [inline sdk/api_key.go:126]: instead of using 404, please use http.StatusNotFound

**Comment by timja** [comment]
> > @Trois-Six I partially reverted one of your fixes for 2fa ([80d523e](https://github.com/Trois-Six/terraform-provider-sendgrid/commit/80d523e88b92eb63efb3acea059b9d21128ef66d))
> > I'm using an account via a reseller (Azure) and adding that scope causes send grid to respond with an internal server error.
> > I'm guessing it didn't work for you without adding the scope? Or any other reason you force added it
> 
> It's not that it doesn't work, but that scope is automatically added even if I do not set it. If I don't do that, each time I run terraform I have a change. It's the same for "sender_verification_eligible".

k I think you need to add it then for your account type, unfortunately not all account types work for it

**Comment by timja** [comment]
Something wrong in the release again?

```
Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
trois-six/sendgrid: no available releases match the given constraints 0.1.5
```
?

**Pull request merged**